## `count_codons.py`
Count how many of each codon is coded for in a DNA file.

### `SARS-CoV-2_separate_genes.fna`
Input to `count_codons.py`. File contains many separate genes

### `SARS-CoV-2_separate_genes_output.csv`
Output to `count_codons.py` when `SARS-CoV-2_separate_genes.fna` is input

### `SARS-CoV-2_whole_genome.fna`
Input to `count_codons.py`. Doesn't separate the file into genes.

### `SARS-CoV-2_whole_genome.csv`
Output to `count_codons.py` when `SARS-CoV-2_whole_genome.fna` is input

### `test_genome.fna`
Input to `count_codons.py`. Used for testing

### `output.csv`
Output to `count_codons.py` when `test_genome.fna` is input. Used for testing. Generated by code

### `test_output.csv`
Output to `count_codons.py` when `test_genome.fna` is input. Used for testing. Generated by me

## `count_aminos.py`
Counts how many of each amino acid is coded for in a DNA file. 

### `whole_aminos_output.csv`
Output to `count_aminos.py` when `SARS-CoV-2_whole_genome.csv` is input.

### `separate_aminos_output.csv`
Output to `count_aminos.py` when `SARS-CoV-2_separate_genes.csv` is input.

## `make_barplot.py`
Plots a histogram of the side by side comparison of two different codon/amino acid counts

### `amino_acid_barplot.png`
Ouput to `make_barplot.py` when two amino acid counts are input

### `codon_barplot.png`
Ouput to `make_barplot.py` when two codon counts are input
